File: src/listeners/actions/block_helpers.py
from slack_sdk import WebClient
from logging import Logger
from datetime import datetime

def assemble_pr_message_blocks(client: WebClient, pr: dict, user_id: str, logger: Logger) -> list:
    try:
        blocks = create_pr_message_block(client, pr, logger)
        logger.info(f"Assembling PR message blocks for PR {pr['id']} by user {user_id}")

        # Append the necessary buttons
        blocks[1]["elements"].append(create_plus_one_button(pr["id"], user_id, pr))
        blocks[1]["elements"].append(create_question_mark_button(pr["id"], user_id, pr))
        blocks[1]["elements"].append(create_remove_button(pr["id"]))
        blocks[1]["elements"].append(create_edit_button(pr["id"]))

        return blocks
    except Exception as e:
        logger.error(f"Error assembling message blocks for PR {pr['id']}: {e}")
        raise

def create_pr_message_block(client: WebClient, pr: dict, logger: Logger) -> list:
    try:
        # Get the submitter's Slack user mention and real name
        submitter_info = client.users_info(user=pr['submitter_id'])
        submitter_name = submitter_info["user"]["real_name"]
        submission_time = datetime.fromisoformat(pr['timestamp']).strftime('%b %-d, %-I:%M %p')

        # Create the list of reviewers who have +1'ed the PR
        reviewers_text = ", ".join(
            [f"*{client.users_info(user=r)['user']['real_name']}*" for r in pr["reviewers"]]
        )
        reviews_needed = pr["reviews_needed"] - pr["reviews_received"]

        # Construct the message with enhanced styling
        message_text = (
            f":memo: *PR Reviews Requested!*\n\n"  # Heading with an icon
            f"*<{pr['link']}|{pr['name']}>*\n"
            f"{pr['description']}\n\n"
            f":bust_in_silhouette: *Submitted by:* *{submitter_name}* at {submission_time}\n"
        )
        
        if reviews_needed <= 0:
            submitter_mention = f"<@{pr['submitter_id']}>"
            message_text += (
                f"\n\n:white_check_mark: Your PR was reviewed by {reviewers_text}!\n"
                f"It’s ready to be merged.\n"
                f"Please merge the PR and remove it from the queue, or update the request if needed."
            )
        else:
            message_text += (
                f"\n:hourglass_flowing_sand: Needs {reviews_needed} more reviews\n\n"
                f":white_check_mark: *+1s:* {reviewers_text}"
            )

        # Add Attention requested line
        if pr["attention_requests"]:
            attention_requested_by = ", ".join(
                [f"*{client.users_info(user=user_id)['user']['real_name']}*" for user_id in pr["attention_requests"]]
            )
            message_text += f"\n\n:speech_balloon: *Attention requested by:* {attention_requested_by}"

        return [
            {
                "type": "section",
                "text": {
                    "type": "mrkdwn",
                    "text": message_text,
                },
            },
            {
                "type": "actions",
                "elements": [],  # Start with an empty list for elements
            }
        ]
    except Exception as e:
        logger.error(f"Error creating PR message block for PR {pr['id']}: {e}")
        raise

def create_plus_one_button(pr_id: str, user_id: str, pr: dict) -> dict:
    return {
        "type": "button",
        "text": {
            "type": "plain_text",
            "text": "+1",
            "emoji": True
        },
        "action_id": f"plus_one_{pr_id}",
        "value": pr_id,
    }

def create_remove_button(pr_id: str) -> dict:
    return {
        "type": "button",
        "text": {
            "type": "plain_text",
            "text": "Remove",
        },
        "action_id": f"remove_pr_{pr_id}",
        "value": pr_id,
        "confirm": {
        "title": {
            "type": "plain_text",
            "text": "Are you sure?"
        },
        "text": {
            "type": "mrkdwn",
            "text": "This action cannot be undone."
        },
        "confirm": {
            "type": "plain_text",
            "text": "Yes, remove it"
        },
        "deny": {
            "type": "plain_text",
            "text": "Cancel"
        }
        }
    }

def create_edit_button(pr_id: str) -> dict:
    return {
        "type": "button",
        "text": {
            "type": "plain_text",
            "text": "Edit",
        },
        "action_id": f"edit_pr_{pr_id}",
        "value": pr_id,
    }

def create_question_mark_button(pr_id: str, user_id: str, pr: dict) -> dict:
    question_mark_button = {
        "type": "button",
        "text": {
            "type": "plain_text",
            "text": ":speech_balloon:",
            "emoji": True
        },
        "action_id": f"attention_request_{pr_id}",
        "value": pr_id,
    }
    if pr["attention_requests"]:
        question_mark_button["style"] = "danger"

    return question_mark_button

File: src/listeners/actions/BlockHelpers.java
import com.slack.api.Slack;
import com.slack.api.methods.SlackApiException;
import com.slack.api.methods.response.users.UsersInfoResponse;
import com.slack.api.model.block.Blocks;
import com.slack.api.model.block.LayoutBlock;
import com.slack.api.model.block.element.BlockElement;
import com.slack.api.model.block.element.BlockElements;
import com.slack.api.model.block.element.ButtonElement;
import org.slf4j.Logger;

import java.io.IOException;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class BlockHelpers {

    public static List<LayoutBlock> assemblePrMessageBlocks(Slack client, Map<String, Object> pr, String userId, Logger logger) {
        try {
            List<LayoutBlock> blocks = createPrMessageBlock(client, pr, logger);
            logger.info("Assembling PR message blocks for PR {} by user {}", pr.get("id"), userId);

            // Append the necessary buttons
            blocks.get(1).getElements().add(createPlusOneButton((String) pr.get("id"), userId, pr));
            blocks.get(1).getElements().add(createQuestionMarkButton((String) pr.get("id"), userId, pr));
            blocks.get(1).getElements().add(createRemoveButton((String) pr.get("id")));
            blocks.get(1).getElements().add(createEditButton((String) pr.get("id")));

            return blocks;
        } catch (Exception e) {
            logger.error("Error assembling message blocks for PR {}: {}", pr.get("id"), e.getMessage());
            throw new RuntimeException(e);
        }
    }

    public static List<LayoutBlock> createPrMessageBlock(Slack client, Map<String, Object> pr, Logger logger) {
        try {
            UsersInfoResponse submitterInfo = client.methods().usersInfo(r -> r.user((String) pr.get("submitter_id")));
            String submitterName = submitterInfo.getUser().getRealName();
            String submissionTime = LocalDateTime.parse((String) pr.get("timestamp"))
                    .format(DateTimeFormatter.ofPattern("MMM d, h:mm a"));

            List<String> reviewersText = new ArrayList<>();
            for (String reviewerId : (List<String>) pr.get("reviewers")) {
                UsersInfoResponse reviewerInfo = client.methods().usersInfo(r -> r.user(reviewerId));
                reviewersText.add("*" + reviewerInfo.getUser().getRealName() + "*");
            }
            String reviewsNeeded = String.valueOf((int) pr.get("reviews_needed") - (int) pr.get("reviews_received"));

            String messageText = ":memo: *PR Reviews Requested!*\n\n" +
                    "*" + pr.get("link") + "|" + pr.get("name") + "*\n" +
                    pr.get("description") + "\n\n" +
                    ":bust_in_silhouette: *Submitted by:* *" + submitterName + "* at " + submissionTime + "\n";

            if (Integer.parseInt(reviewsNeeded) <= 0) {
                messageText += "\n\n:white_check_mark: Your PR was reviewed by " + String.join(", ", reviewersText) +
                        "!\nIt’s ready to be merged.\nPlease merge the PR and remove it from the queue, or update the request if needed.";
            } else {
                messageText += "\n:hourglass_flowing_sand: Needs " + reviewsNeeded + " more reviews\n\n" +
                        ":white_check_mark: *+1s:* " + String.join(", ", reviewersText);
            }

            if (!((List<String>) pr.get("attention_requests")).isEmpty()) {
                List<String> attentionRequestedBy = new ArrayList<>();
                for (String userId : (List<String>) pr.get("attention_requests")) {
                    UsersInfoResponse attentionRequesterInfo = client.methods().usersInfo(r -> r.user(userId));
                    attentionRequestedBy.add("*" + attentionRequesterInfo.getUser().getRealName() + "*");
                }
                messageText += "\n\n:speech_balloon: *Attention requested by:* " + String.join(", ", attentionRequestedBy);
            }

            return List.of(
                    Blocks.section(section -> section.text(text -> text.type("mrkdwn").text(messageText))),
                    Blocks.actions(actions -> actions.elements(new ArrayList<>())) // Placeholder for buttons
            );
        } catch (IOException | SlackApiException e) {
            logger.error("Error creating PR message block for PR {}: {}", pr.get("id"), e.getMessage());
            throw new RuntimeException(e);
        }
    }

    public static BlockElement createPlusOneButton(String prId, String userId, Map<String, Object> pr) {
        return BlockElements.button(b -> b
                .text(pt -> pt.type("plain_text").text("+1").emoji(true))
                .actionId("plus_one_" + prId)
                .value(prId)
        );
    }

    public static BlockElement createRemoveButton(String prId) {
        return BlockElements.button(b -> b
                .text(pt -> pt.type("plain_text").text("Remove"))
                .actionId("remove_pr_" + prId)
                .value(prId)
                .confirm(confirm -> confirm
                        .title(title -> title.type("plain_text").text("Are you sure?"))
                        .text(text -> text.type("mrkdwn").text("This action cannot be undone."))
                        .confirm(c -> c.type("plain_text").text("Yes, remove it"))
                        .deny(d -> d.type("plain_text").text("Cancel"))
                )
        );
    }

    public static BlockElement createEditButton(String prId) {
        return BlockElements.button(b -> b
                .text(pt -> pt.type("plain_text").text("Edit"))
                .actionId("edit_pr_" + prId)
                .value(prId)
        );
    }

    public static BlockElement createQuestionMarkButton(String prId, String userId, Map<String, Object> pr) {
        ButtonElement questionMarkButton = BlockElements.button(b -> b
                .text(pt -> pt.type("plain_text").text(":speech_balloon:").emoji(true))
                .actionId("attention_request_" + prId)
                .value(prId)
        );
        if (!((List<String>) pr.get("attention_requests")).isEmpty()) {
            questionMarkButton.setStyle("danger");
        }
        return questionMarkButton;
    }
}
